### PR TITLE
Add public API definitions for surfacing data age

### PR DESCRIPTION
**Pull request by jowlyzhang**
This PR adds the definition for the public APIs for surfacing data write time info. It only contains minimum implementation. The implementations will be in follow ups. I need to sync with customers if these public APIs meet their requirements and are easy to use. And make modifications accordingly before proceeding with implementations.

- `struct DataCollectionUnixWriteTimeInfo` is a struct for the unix write time info for a collection of data
- `DB::GetPropertiesOfTablesForLevels` returns table properties collection per level
- `GetDataCollectionUnixWriteTimeInfoForFile` returns the data write time info for a file.
- `GetDataCollectionUnixWriteTimeInfoForLevels` returns the data write time info for levels.
- The user property names for recording write time stats in the user collected properties are defined.

Follow ups:
- Implement collecting the write time related user table properties
- Use the data write time info recorded in the table properties to implement these APIs

Test plan:
No functional change, also follow ups should have tests covering the minimum implementation added in this PR.

### CODE REVIEW

File: include/rocksdb/utilities/table_properties_collectors.h [hunk 1]
```diff
@@ -130,4 +132,81 @@ class CompactForTieringCollectorFactory
 
 std::shared_ptr<CompactForTieringCollectorFactory>
 NewCompactForTieringCollectorFactory(double compaction_trigger_ratio);
+
+// Information for the age of a collection of data. The unit for all the age
+// stats are in seconds. Data's age is defined as the time passed since the data
+// was initially written into the DB. Check `DataCollectionIsEmpty` and
```
Comment (pdillinger) [inline include/rocksdb/utilities/table_properties_collectors.h:138]: Does this mean the structure becomes less accurate over time, since data with age 1000 seconds needs to be 2000 seconds in another 1000 seconds? Any reason not to use unix times?

File: include/rocksdb/db.h [hunk 2]
```diff
@@ -2014,6 +2015,19 @@ class DB {
       ColumnFamilyHandle* column_family, const Range* range, std::size_t n,
       TablePropertiesCollection* props) = 0;
 
+  // Get the aggregated data age info per level. Data age info is only available
+  // when DB's time tracking is enabled and data's age info tracking is enabled.
+  // The former is enabled with `preserve_internal_time_seconds` and/or
+  // `preclude_last_level_data_seconds`. The latter is enabled via
+  // `CompactForTieringCollectorFactory`.
+  virtual Status GetDataCollectionAgeInfoForLevels(
```
Comment (pdillinger) [inline include/rocksdb/db.h:2023]: This seems a little niche to go in DB API. If we had a version of GetPropertiesOfAllTables that divided the data by level (rather than cross-refing with GetLiveFilesMetaData) then collating the data could be a utility function outside of DB.

Comment (jowlyzhang) [reply include/rocksdb/db.h:2023]: That's fair. I had a similar version that divide the functionality into a table property collecting API and a util API to convert that into data age. I was thinking not many other usage expect table properties per level. But that is a more general API. I think Zippy folks should be fine with it too. Thanks for the suggestion!

File: include/rocksdb/utilities/table_properties_collectors.h (hunk 1 above)
Comment (pdillinger) [reply include/rocksdb/utilities/table_properties_collectors.h:138]: My worry is that the struct makes it a little too easy to "do the wrong thing" such as saving the information for minutes or hours or days and forgetting that it is out of date. I would be satisfied if the struct contains a field for the unix time that the ages are relative to. Or maybe it's better for the struct to store the data as unix times and expose the ages through public functions, perhaps relative to a specified "current time."

I see this as somewhat analogous to my argument for using UTC time in `daily_offpeak_time_utc`. The simplest, most "convenient" API is usually best but not if it invites bugs. Prompting the user to think about the potential bug-prone piece (what is the reference point for these times?) is the best approach IMHO.

> I figured this is a more intuitive definition of data's age, if we directly use unix time, I'm not sure the definition of averaging this unix time age is as meaningful.

(Assuming exclusion of "infinitely old"...) Yeah, it may not be obvious to general users that average/min/max absolute times is equivalent to average/min/max relative time (even though these are robust under linear transformations and then the reverse transformation)

Comment (jowlyzhang) [reply include/rocksdb/utilities/table_properties_collectors.h]: Thanks for the very detailed context! Yeah, unix time is really standard practices. The seemingly convenient `age` API I was trying to provide is really unnecessary and dangerous design as you pointed out. Users can easily do a (now() - write_time) query to get age themselves. I should have been a non-general user myself to understand that absolute time and relative time work the same for these statistics, thanks for calling that out!

**Review by pdillinger — APPROVED** [review]
LGTM, thanks!
